app/commands/generate_missing_mappings_command.rb

```diff
@@ -1,68 +1,116 @@
     end
   end
 
   private
 
+  def setup_options
+    params[:shopify_version] ||= sys.read_file("VERSION").strip
+  end
+
   def find_unmapped_categories
-    spinner("Searching Shopify categories that lack mappings") do
-      @unmapped_category_groups = []
-      all_shopify_category_ids = Set.new(Category.all.pluck(:id))
-      latest_shopify_version = "shopify/#{sys.read_file("VERSION").strip}"
-      MappingRule.where(
-        input_version: latest_shopify_version,
-      ).group_by(&:output_version).each do |output_version, mappings|
-        shopify_category_ids_from_mappings_input = Set.new(
-          mappings.map do |mapping|
-            mapping.input.product_category_id.split("/").last
-          end,
-        )
-        unmapped_category_ids = all_shopify_category_ids - shopify_category_ids_from_mappings_input
-        category_ids_full_names = unmapped_category_ids.sort.map do |id|
-          category_full_name = Category.find(id)&.full_name
-          [id, category_full_name] if category_full_name
-        end.compact.to_h
+    spinner("Finding Shopify categories that are unmapped") do
+      all_shopify_ids = Set.new(Category.all.pluck(:id))
+      mappings_by_output = MappingRule
+        .where(input_version: "shopify/#{params[:shopify_version]}")
+        .group_by(&:output_version)
+      @unmapped_categories = mappings_by_output.filter_map do |output_taxonomy, mappings|
+        # Can this be an ActiveRecord call? Should it?
+        # TODO: Where is #product_category_id defined? AFAICT #input returns a hash
+        found_shopify_ids = Set.new(mappings.map { _1.input.product_category_id.split("/").last })
+        unmapped_ids = all_shopify_ids - found_shopify_ids
+        category_ids_full_names = Category.where(id: unmapped_ids).pluck(:id, :full_name).to_h
         next if category_ids_full_names.empty?
 
-        @unmapped_category_groups << {
-          input_taxonomy: mappings.first.input_version,
-          output_taxonomy: output_version,
-          category_ids_full_names: category_ids_full_names,
-        }
+        { output_taxonomy:, category_ids_full_names: }
+      end
+    end
+  end
+
+  def generate_missing_mappings
+    frame("Generating missing mappings") do
+      @unmapped_categories.each do |unmapped_category_group|
+        output_taxonomy = unmapped_category_group[:output_taxonomy]
+        index_name = output_taxonomy.gsub(%r{[/\-]}, "_")
+
+        frame("Generating mappings to #{output_taxonomy}") do
+          embedding_data = load_embedding_data(output_taxonomy)
+          index_embedding_data(embedding_data:, index_name:)
+          generate_and_evaluate_mappings_for_group(
+            unmapped_category_group:,
+            index_name:,
+          )
+        end
       end
     end
   end
 
   def load_embedding_data(output_taxonomy)
-    embeddings = nil
-    spinner("Loading embeddings for #{output_taxonomy}") do
+    spinner("Loading embeddings for #{output_taxonomy}") do |sp|
```

Comment:
spinners return the results of their block; lets us save the var passing

Reply:
Also given the frame, we can probably drop `for #{output_taxonomy}` for everything called within said frame